mm/kmap/kmm_map.c

```diff
@@ -0,0 +1,466 @@
+
+  vaddr = (uintptr_t)map_pages(pages, npages, prot);
+  if (!vaddr)
+    {
+      return NULL;
+    }
+
+  return (FAR void *)(vaddr);
+}
+
+/****************************************************************************
+ * Name: kmm_unmap
+ *
+ * Description:
+ *   Unmap a previously allocated kernel virtual memory area.
+ *
+ * Input Parameters:
+ *   kaddr - The kernel virtual address where the mapping begins.
+ *
+ * Returned Value:
+ *   None.
+ *
+ ****************************************************************************/
+
+void kmm_unmap(FAR void *kaddr)
+{
+  FAR struct mm_map_entry_s *entry;
+  unsigned int               npages;
+  int                        ret;
+
+  /* Lock the mapping list when we fiddle around with it */
+
+  ret = kmm_map_lock();
+  if (ret == OK)
+    {
+      /* Find the entry, it is OK if none found */
+
+      entry = mm_map_find(get_current_mm(), kaddr, 1);
+      if (entry)
+        {
+          npages = MM_NPAGES(entry->length);
+
+          /* Remove the mappings from the page directory */
+
+          up_addrenv_kunmap_pages((uintptr_t)entry->vaddr, npages);
+
+          /* Release the virtual memory area for use */
+
+          gran_free(&g_kmm_map_vpages, entry->vaddr, entry->length);
+
+          /* Remove the mapping from the kernel mapping list */
+
+          mm_map_remove(&g_kmm_map, entry);
+        }
+
+      kmm_map_unlock();
+    }
+}
+
+/****************************************************************************
+ * Name: kmm_user_map
+ *
+ * Description:
+ *   Map a region of user memory (physical pages) for kernel use through
+ *   a continuous virtual memory area.
+ *
+ * Input Parameters:
+ *   uaddr - The user virtual address where mapping begins.
+ *   size  - Size of the region.
+ *
+ * Returned Value:
+ *   Pointer to the virtual memory area, or NULL if out of memory.
+ *
+ ****************************************************************************/
+
+FAR void *kmm_user_map(FAR void *uaddr, size_t size)
+{
+  FAR void *pages;
+  uintptr_t vaddr;
+  uintptr_t offset;
+  size_t    npages;
+  int       ret;
+
+  /* Find the kernel addressable virtual address, or map the user memory */
+
+  vaddr = (uintptr_t)uaddr;
+
+  /* If the memory is not user memory, get out */
+
+  if (!up_addrenv_user_vaddr(vaddr))
+    {
+      /* Not user memory, get out */
+
+      return uaddr;
+    }
+
+  /* How many pages (including partial) does the space encompass ? */
+
+  offset = vaddr & MM_PGMASK;
+  vaddr  = MM_PGALIGNDOWN(vaddr);
+  npages = MM_NPAGES(offset + size);
+
+  /* Does the area fit in 1 page, including page boundary crossings ? */
+
+  if (npages == 1)
+    {
+      /* Yes, can simply return the kernel addressable virtual address */
+
+      vaddr = (uintptr_t)map_single_user_page(vaddr);
+      return (FAR void *)(vaddr + offset);
+    }
+
+  /* No, the area must be mapped into kernel virtual address space */
+
+  pages = kmm_zalloc(npages * sizeof(void *));
```

Comment:
`sizeof(FAR void *)`

Reply:
Unnecessary, the size of void* does not change if declared far

Reply:
Just for style consistency.

Reply:
Yeah, just rechecked. 16 bits for near and 32 bits for far